lib/spack/spack/installer.py

```diff
@@ -1489,11 +1520,9 @@ def install(self, **kwargs):
             # Skip the installation if the spec is not being installed locally
             # (i.e., if external or upstream) BUT flag it as installed since
             # some package likely depends on it.
-            if pkg_id != self.pkg_id:
-                not_local = _handle_external_and_upstream(pkg, False)
-                if not_local:
-                    self._update_installed(task)
-                    _print_installed_pkg(pkg.prefix)
+            if not task.explicit:
+                if _handle_external_and_upstream(pkg, False):
+                    self._flag_installed(pkg, get_dependent_ids(spec))
```

Comment:
I had some issues when, e.g., `cmake` showed up several times in separately concretized specs, and it was not removed from all dependents' uninstalled dependencies. This fixed said scenario with my small reproducer (using our internal packages, unfortunately):
```suggestion
                    self._flag_installed(pkg, task.dependents)
```
At least here, I think it's better to flag for the superset of dependents (which is also readily available), rather than just the ones from one of the specs in the tree.

Reply:
> I had some issues when, e.g., an external `cmake` showed up several times in separately concretized specs, and it was not removed from all dependents' uninstalled dependencies. This fixed said scenario with my small reproducer (using our internal packages, unfortunately):
> 
> At least here, I think it's better to flag for the superset of dependents (which is also readily available), rather than just the ones from one of the specs in the tree.

Good point, thanks!

Reviewing other calls to `_flag_installed` makes me realize there's a small, related refactor here that should make the processing being done here a bit clearer.